agent/uiserver/uiserver.go

```diff
@@ -117,6 +129,61 @@ func (h *Handler) getState() *reloadableState {
 	return nil
 }
 
+func (h *Handler) serveUIMetricsProviders(resp http.ResponseWriter, req *http.Request) {
+	// Reload config in case it's changed
+	state := h.getState()
+
+	if len(state.cfg.MetricsProviderFiles) < 1 {
+		http.Error(resp, "No provider JS files configured", http.StatusNotFound)
+		return
+	}
+
+	var buf bytes.Buffer
+
+	// Open each one and concatenate them
+	for _, file := range state.cfg.MetricsProviderFiles {
+		base := path.Base(file)
+		_, err := buf.WriteString("// " + base + "\n\n")
+		if err != nil {
+			http.Error(resp, "Internal Server Error", http.StatusInternalServerError)
+			h.logger.Error("failed writing provider JS files: %s", err)
+			return
+		}
+
+		// Attempt to open the file
+		f, err := os.Open(file)
```

Comment:
I believe this file is never closed. It would probably have a `defer f.Close()`, right?  Since this is in a loop, I guess we need to extract most of this out into a new function so that the defer happens before the next iteration of the loop.

A separate function should also simplify this error handling by returning an error and allowing us to only write the error response in one place.

Reply:
Separate func makes this way cleaner thanks. Great suggestion.